pinot-controller/src/main/java/com/linkedin/pinot/controller/ControllerLeadershipManager.java

```diff
@@ -0,0 +1,116 @@
+/**
+ * Copyright (C) 2014-2018 LinkedIn Corp. (pinot-core@linkedin.com)
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *         http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.linkedin.pinot.controller;
+
+import java.util.Map;
+import java.util.concurrent.ConcurrentHashMap;
+import org.apache.helix.HelixManager;
+import org.apache.helix.api.listeners.ControllerChangeListener;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+
+/**
+ * Single place for listening on controller changes
+ * This should be created at controller startup and everyone who wants to listen to controller changes should subscribe
+ */
+public class ControllerLeadershipManager {
+
+  private static final Logger LOGGER = LoggerFactory.getLogger(ControllerLeadershipManager.class);
+
+  private static ControllerLeadershipManager INSTANCE = null;
+
+  private HelixManager _helixManager;
+  private volatile boolean _amILeader = false;
+
+  private Map<String, ControllerChangeSubscriber> _subscribers = new ConcurrentHashMap<>();
+
+  private ControllerLeadershipManager(HelixManager helixManager) {
+    _helixManager = helixManager;
+    _helixManager.addControllerListener((ControllerChangeListener) notificationContext -> onControllerChange());
+  }
+
+  /**
+   * Create an instance of ControllerLeadershipManager
+   * @param helixManager
+   */
+  public static synchronized void createInstance(HelixManager helixManager) {
```

Comment:
It might be simpler to expose just getInstance(); if its not created, the instance will get created. Does the user code really need to distinguish between create and get?

Reply:
I thought about that as well. I think the option is between create() and init(HelixManager). Perhaps the latter is a bit less confusing.

Reply:
We need 2 calls, else everyone who wants to get an instance will have to provide helix manager

Reply:
I think init() and getInstance() is less confusing and less prone to errors than create() and getInstance().